Feature: implement `VADTree` plugin.

digitalisx opened the pull request:
## Description

Hello, everyone in the community! 😃
There are some plugins that have not been implemented as they are updated from [Volatility2](https://github.com/volatilityfoundation/volatility/wiki/Command-Reference#vadtree) to 3.
I found that `VADTree` plugin has not yet migrated to 3.
So I'm implemented (or porting) of `VADTree` plugin according to the Volatility3 structure.

## Command
Help Command

```
> python3 vol.py -h
windows.vadtree.VadTree Walk the VAD tree and display in tree format.
```

Run Command

```shell
> python3 vol.py -f case.vmem windows.vadtree
```

## Result

```shell
> python3 vol.py -f case.vmem -r pretty windows.vadtree --pid=508 
Volatility 3 Framework 2.2.0
Formatting...0.00               PDB scanning finished                        
         | PID |   Process |         Offset | Type |          Start |            End |  Tag
*        | 508 | csrss.exe | 0x97065bf58dd0 |  N/A |  0x1842f7a0000 |  0x1842f7a0fff | VadS
**       | 508 | csrss.exe | 0x97065bf575c0 |  N/A |  0x1842da00000 |  0x1842dbfffff | VadS
***      | 508 | csrss.exe | 0x97065bf58b00 |  N/A |   0x6086fc0000 |   0x6086ffffff | VadS
****     | 508 | csrss.exe | 0x97065bf57520 |  N/A |   0x6086e80000 |   0x6086ebffff | VadS
*****    | 508 | csrss.exe | 0x97065b6771f0 |  N/A |     0x7ffe2000 |     0x7ffe2fff | VadS
******   | 508 | csrss.exe | 0x97065b6777e0 |  N/A |     0x7ffe0000 |     0x7ffe0fff | VadS
******   | 508 | csrss.exe | 0x97065b677650 |  N/A |   0x6086c00000 |   0x6086dfffff | VadS
*******  | 508 | csrss.exe | 0x97065bf58ec0 |  N/A |   0x6086ba0000 |   0x6086bdffff | VadS
*****    | 508 | csrss.exe | 0x97065bf5cbb0 |  N/A |   0x6086f00000 |   0x6086f3ffff | VadS
******   | 508 | csrss.exe | 0x97065bf58ab0 |  N/A |   0x6086ec0000 |   0x6086efffff | VadS
******   | 508 | csrss.exe | 0x97065bf58a60 |  N/A |   0x6086f80000 |   0x6086fbffff | VadS
****     | 508 | csrss.exe | 0x97065bf57840 | Heap |  0x1842d8c0000 |  0x1842d8cafff | VadS
*****    | 508 | csrss.exe | 0x97065b481a60 | File |  0x1842d880000 |  0x1842d880fff | Vad 
******   | 508 | csrss.exe | 0x97065bf58f60 |  N/A |   0x6087040000 |   0x608707ffff | VadS
*******  | 508 | csrss.exe | 0x97065bf58d30 |  N/A |   0x6087000000 |   0x608703ffff | VadS
*******  | 508 | csrss.exe | 0x97065bf59fa0 |  N/A |   0x6087080000 |   0x60870bffff | VadS
```

However, I implementing logic for decision VAD Type, so I leave it as a draft PR.

digitalisx committed:
Add: vadtree plugin initialize version

digitalisx committed:
Add: get heaps, modules method

digitalisx committed:
Add: get threads method

digitalisx committed:
Merge branch 'volatilityfoundation:develop' into feature/vadtree

digitalisx committed:
Add: get stacks, types method

digitalisx committed:
Merge branch 'feature/vadtree' of https://github.com/digitalisx/volatility3 into feature/vadtree

digitalisx committed:
Add: EOF

digitalisx committed:
Merge branch 'volatilityfoundation:develop' into feature/vadtree

digitalisx committed:
Merge branch 'volatilityfoundation:develop' into feature/vadtree

digitalisx committed:
Merge branch 'volatilityfoundation:develop' into feature/vadtree

digitalisx committed:
Merge branch 'volatilityfoundation:develop' into feature/vadtree

ikelos commented on volatility3/framework/plugins/windows/vadtree.py:
```diff
@@ -0,0 +1,143 @@
+            levels = {}
+
+            for vad in vadinfo.VadInfo.list_vads(proc):
+                level = levels.get(vad.get_parent(), -1) + 1
```
So, this line and the next should probably be using `& self.context.layers[vad.vol.layer_name].address_mask` to ensure consistency.  This can also be used for the output in vadwalk to produce similar looking results.

digitalisx replied:
From what I understand, should I think that implementing `VADWalk` and `VADTree`, which are consistent through `address_mask`, is a good direction? :)
Then, I think we do not calculate the offset of VAD strangely.

digitalisx replied:
After testing it, I think it's working well! Thank you for your help.
If there is no problem, I will close #731

digitalisx committed:
Fix: address mask from @ikelos help

digitalisx commented:
Added a description of the PR that was quickly submitted to draft to handle this issue (#731).
This is still included in my interest and work object. 🙂

digitalisx committed:
Merge branch 'volatilityfoundation:develop' into feature/vadtree

digitalisx committed:
Merge remote-tracking branch 'origin/develop' into feature/vadtree

digitalisx committed:
Update VadTree plugin to current framework APIs

In the four years since this PR was opened, the framework's plugin-
dependency declaration and the windows.pslist API have changed:

  * requirements.PluginRequirement was retired in favour of the more
    general requirements.VersionRequirement, which takes 'component'
    rather than 'plugin' as the keyword for the satisfying class;
  * windows.pslist.PsList.list_processes() now takes a single
    'kernel_module_name' argument instead of the (layer_name,
    symbol_table) pair that VadTree was supplying;
  * windows.pslist.PsList is now version (3, 0, 0).

The 'modules' plugin dependency declared in get_requirements() was
unused (the file only referenced 'modules' as a local variable inside
get_modules()), so the import and the requirement are dropped together.

digitalisx committed:
Fix get_heaps to enumerate every PEB heap, not just the first

proc.get_peb().ProcessHeaps is a pointer to an array of pointers whose
length is recorded in PEB.NumberOfHeaps. The previous implementation
called .dereference() on that pointer, which returns only the first
entry. Consequently, VAD type classification could mark just the
default process heap as "Heap" and every other heap allocation showed
up as "N/A".

The fix instantiates the array explicitly using NumberOfHeaps and the
pointer type from the matching symbol table, and returns every non-null
heap base. The method is also converted from a classmethod to a regular
method so it can use self.context to construct the array.

digitalisx committed:
Cache heaps/modules/stacks once per process

The previous implementation called get_heaps(), get_modules() and
get_stacks() from inside get_type(), which itself ran once for every
VAD. For a process with N VADs that produced N * (heap_walk +
module_walk + thread_walk) operations -- noticeably slow on processes
with hundreds or thousands of VADs.

Hoist the three lookups into _generator, evaluate them exactly once
per process, and pass the resulting address sets into get_type. Using
frozensets also turns the inner membership tests from O(K) list scans
into O(1) hash lookups.

digitalisx committed:
Stop shadowing the type() builtin and finish dead-code cleanup

The classifier used 'type' as its local variable, which shadowed the
Python builtin and made the function harder to read or extend. Rename
the local to 'vad_type' everywhere, including the matching variable
in _generator that receives its return value.

While here, replace the leftover commented-out 'vollog.log()' lines
with actual vollog.debug calls -- vollog was imported but otherwise
unused -- and fill in the empty docstrings for get_modules and
get_type. Drop the unnecessary parentheses around 'if vad:' and stop
recomputing vad.get_start() three times in a row inside get_type.
